replace uses of pkg_resources

we got
```
FAILED ../core/local/tests/test_create.py::test_create_withcfg - DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
FAILED ../core/local/tests/test_create.py::test_bad_cfg_proc - DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
FAILED ../local/tests/test_addurls.py::TestAddurls::test_addurls_create_newdataset - DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
FAILED ../local/tests/test_addurls.py::TestAddurls::test_addurls_subdataset - DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
FAILED ../local/tests/test_run_procedure.py::test_run_proc_with_dict - DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
FAILED ../resources/procedures/tests/test_noannex.py::test_noannex_simple - DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
FAILED ../resources/procedures/tests/test_noannex.py::test_noannex_create_force - DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
FAILED ../local/tests/test_run_procedure.py::test_invalid_call - DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
FAILED ../local/tests/test_run_procedure.py::test_dirty - DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
FAILED ../local/tests/test_run_procedure.py::test_procedure_discovery - DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
FAILED ../local/tests/test_run_procedure.py::test_configs - DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
FAILED ../local/tests/test_run_procedure.py::test_spaces - DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
FAILED ../local/tests/test_run_procedure.py::test_text2git - DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
```
for which I will workaround by ignoring it for now, but should be addressed properly following the recommendations.

--- a/changelog.d/pr-7439.md
+++ b/changelog.d/pr-7439.md
@@ -1,0 +1,3 @@
+### 🏠 Internal
+
+- Eliminate uses of `pkg_resources`.  Fixes [#7435](https://github.com/datalad/datalad/issues/7435) via [PR #7439](https://github.com/datalad/datalad/pull/7439) (by [@jwodder](https://github.com/jwodder))

--- a/datalad/cli/parser.py
+++ b/datalad/cli/parser.py
@@ -11,29 +11,29 @@ It is unconditionally imported by the main() entrypoint.
 # like error handling, must be done conditionally in-line.
 
 import argparse
+import logging
+import sys
 from collections import defaultdict
 from functools import partial
-import sys
-
 
 from datalad import __version__
-
-from .common_args import common_args
 from datalad.interface.base import (
-    is_api_arg,
     get_cmd_doc,
     get_interface_groups,
+    is_api_arg,
     load_interface,
 )
+from datalad.support.constraints import EnsureChoice
 from datalad.utils import getargspec
+
+from .common_args import common_args
+from .exec import call_from_parser
+from .helpers import get_commands_from_groups
 from .interface import (
     alter_interface_docs_for_cmdline,
     get_cmd_ex,
     get_cmdline_command_name,
 )
-from datalad.support.constraints import EnsureChoice
-from .helpers import get_commands_from_groups
-from .exec import call_from_parser
 
 # special case imports
 #  .helpers import add_entrypoints_to_interface_groups
@@ -43,7 +43,6 @@ from .exec import call_from_parser
 #  .interface._known_extension_commands
 #  .interface._deprecated_commands
 
-import logging
 lgr = logging.getLogger('datalad.cli.parser')
 
 
@@ -138,6 +137,7 @@ def setup_parser(
         # we need the full help, or we have a potential command that
         # lives in an extension, must load all extension, expensive
         from .helpers import add_entrypoints_to_interface_groups
+
         # need to load all the extensions and try again
         # TODO load extensions one-by-one and stop when a command was found
         add_entrypoints_to_interface_groups(interface_groups)
@@ -404,8 +404,8 @@ def try_suggest_extension_with_command(parser, cmd, completing, known_cmds):
     """If completing=False, this function will trigger sys.exit()"""
     # check if might be coming from known extensions
     from .interface import (
-        _known_extension_commands,
         _deprecated_commands,
+        _known_extension_commands,
     )
     extension_commands = {
         c: e
@@ -541,8 +541,13 @@ def parser_add_version_opt(parser, mod_name, include_name=False, delay=False):
             # Let's use the standard Python mechanism if underlying module
             # did not provide __version__
             try:
-                import pkg_resources
-                version = pkg_resources.get_distribution(mod_name).version
+                if sys.version_info < (3, 10):
+                    import importlib_metadata as im
+                else:
+                    import importlib.metadata as im
+
+                pkg = im.packages_distributions()[mod_name][0]
+                version = im.version(pkg)
             except Exception:
                 version = "unknown"
         if include_name:

--- a/datalad/local/run_procedure.py
+++ b/datalad/local/run_procedure.py
@@ -48,6 +48,17 @@ from datalad.utils import (
     split_cmdline,
 )
 
+if sys.version_info < (3, 9):
+    from importlib_resources import (
+        as_file,
+        files,
+    )
+else:
+    from importlib.resources import (
+        as_file,
+        files,
+    )
+
 lgr = logging.getLogger('datalad.local.run_procedures')
 
 
@@ -148,23 +159,17 @@ def _get_procedure_implementation(name='*', ds=None):
 
     # 3. check extensions for procedure
     from datalad.support.entrypoints import iter_entrypoints
-    # delay heavy import until here
-    from pkg_resources import (
-        resource_filename,
-        resource_isdir,
-    )
+
     for epname, epmodule, _ in iter_entrypoints('datalad.extensions'):
-        # use of '/' here is OK wrt to platform compatibility
-        if resource_isdir(epmodule, 'resources/procedures'):
-            for m, n in _get_file_match(
-                    resource_filename(epmodule, 'resources/procedures'),
-                    name):
-                yield (m, n,) + _get_proc_config(n)
+        res = files(epmodule) / "resources" / "procedures"
+        if res.is_dir():
+            with as_file(res) as p:
+                for m, n in _get_file_match(p, name):
+                    yield (m, n,) + _get_proc_config(n)
     # 4. at last check datalad itself for procedure
-    for m, n in _get_file_match(
-            resource_filename('datalad', 'resources/procedures'),
-            name):
-        yield (m, n,) + _get_proc_config(n)
+    with as_file(files("datalad") / "resources" / "procedures") as p:
+        for m, n in _get_file_match(p, name):
+            yield (m, n,) + _get_proc_config(n)
 
 
 def _guess_exec(script_file):

--- a/datalad/support/external_versions.py
+++ b/datalad/support/external_versions.py
@@ -8,18 +8,19 @@
 # ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
 """Module to help maintain a registry of versions for external modules etc
 """
+import os.path as op
 import re
 import sys
-import os.path as op
+from itertools import chain
 from os import linesep
 
-from itertools import chain
 from looseversion import LooseVersion
 
-from datalad.log import lgr
 # import version helper from config to have only one implementation
 # config needs this to avoid circular imports
 from datalad.config import get_git_version as __get_git_version
+from datalad.log import lgr
+
 from .exceptions import (
     CapturedException,
     CommandError,
@@ -46,14 +47,15 @@ class UnknownVersion:
 # Custom handlers
 #
 from datalad.cmd import (
-    WitlessRunner,
     GitWitlessRunner,
     StdOutErrCapture,
+    WitlessRunner,
 )
 from datalad.support.exceptions import (
     MissingExternalDependency,
     OutdatedExternalDependency,
 )
+
 _runner = WitlessRunner()
 _git_runner = GitWitlessRunner()
 
@@ -156,7 +158,10 @@ def get_rsync_version():
     # that of the debian package it's installed with. Reason is in gh-7320,
     # which results in the need to detect a patched-by-ubuntu version of rsync
     # and therefore the package version, not the result of `rsync --version`.
-    from datalad.utils import on_linux, get_linux_distribution
+    from datalad.utils import (
+        get_linux_distribution,
+        on_linux,
+    )
     if on_linux:
         dist = get_linux_distribution()[0]
         if dist in ['debian', 'ubuntu']:
@@ -239,12 +244,16 @@ class ExternalVersions(object):
                 version = getattr(value, attr)
                 break
 
-        # try pkg_resources
+        # try importlib.metadata
         if version is None and hasattr(value, '__name__'):
             pkg_name = klass._PYTHON_PACKAGES.get(value.__name__, value.__name__)
             try:
-                import pkg_resources
-                version = pkg_resources.get_distribution(pkg_name).version
+                if sys.version_info < (3, 10):
+                    import importlib_metadata as im
+                else:
+                    import importlib.metadata as im
+
+                version = im.version(pkg_name)
             except Exception:
                 pass
 

--- a/setup.py
+++ b/setup.py
@@ -29,6 +29,7 @@ requires = {
         'colorama; platform_system=="Windows"',
         'distro; python_version >= "3.8"',
         'importlib-metadata >=3.6; python_version < "3.10"',
+        'importlib-resources >= 3.0; python_version < "3.9"',
         'iso8601',
         'humanize',
         'fasteners>=0.14',

